### PR TITLE
Optimized core paths and fixed several defects.

### PR DESCRIPTION
**Pull request by AlexSkrypnyk**
## Summary

Performance optimizations across the core scan and compare paths, plus correctness fixes, all preserving the public API. Every change is internal (private/protected bodies, local logic) - no class, interface, method signature, public constant, or public parameter name changed. Two independent analysis passes (one for performance, one for defects) drove the work; each finding was verified against the code before acting.

The branch bundles three performance optimizations and three fixes, each with tests green. It also surfaces four additional defects that are **not** fixed here because they would change the on-disk snapshot format (or an already-tested public contract); those are documented below for a deliberate decision rather than changed autonomously.

## Performance optimizations

All are behavior-preserving and covered by the existing suite. The CI benchmark job compares each against the baseline and comments the measured deltas.

| # | Change | What was wasteful | Benchmarks helped |
|---|---|---|---|
| 1 | `IndexedFile`: memoize `getPathnameFromBasepath()`; replace `str_replace($basepath, '', $path)` with `substr($path, strlen($basepath))` | The relative path was recomputed on every call (scan, then compare/sync/patch), and `str_replace` scanned the whole string despite `str_starts_with` already proving the prefix. `substr` is faster **and** correct - `str_replace` also mis-strips a basepath that recurs later in the path. | all (per-file) |
| 2 | `Index::scan()`: `array_unique()` the cached pattern arrays | `Snapshot::compare()` shares one `Rules` object across both indexes, and each `Index` constructor re-appends the default `.ignorecontent`/`.git/` skip rules, so every file was matched against duplicated patterns. | all compare + sync/patch |
| 3 | `Comparer::compare()`: reuse the index key as the diff key and reset the filter cache once | `addLeftFile()`/`addRightFile()` recomputed `getPathnameFromBasepath()` per file and did `$this->cache = []` on **every** call, though the cache is provably empty during `compare()`. The index already keys files by that exact relative path. | all compare |

`addLeftFile()`/`addRightFile()` remain unchanged for external callers - `Comparer::compare()` just stops routing through them redundantly while building the diff set.

## Defects fixed (behavior-preserving, each with a regression test)

**1. `Index::getFiles($cb)` corrupted the cached index.** The callback results were written back into `$this->files`, so a second `getFiles()` (or any later `compare()`) saw transformed values instead of `IndexedFile` objects - a `getFiles(fn)` followed by `compare()` threw a `TypeError`. Now the callback transforms a copy and the cached index is left intact (mirroring `Comparer::filterCached()`).

**2. Rendering a one-sided diff threw an uncatchable `\Error`.** `getAbsentLeftDiffs()`/`getAbsentRightDiffs()` return `Diff` objects with only one side set; calling the public `render()` on one hit `getLeft()`/`getRight()` on an uninitialized typed property and threw `Error` (not a catchable `SnapshotException`). `doRender()` now treats a missing side as empty content, rendering it as a clean addition or deletion.

**3. A lone `!` line in `.ignorecontent` raised a PHP warning.** `parse()` read `$line[1]` unconditionally, so a bare `!` (or `!^`) produced an "Uninitialized string offset" warning - a hard error under PHPUnit's strict handler. The index access is now guarded and a degenerate negation line is ignored.

## Defects found but NOT fixed here (need a format / contract decision)

These are real, reproduced defects. They are left for a deliberate decision because fixing them changes the on-disk snapshot format or an already-tested public contract - not something to do silently in an autonomous PR.

**A. `diff` -> `patch` round-trip corrupts a changed trailing newline (high).** If a file's trailing-newline state differs between baseline and actual, the round-trip silently adds or drops the final newline. `Diff::doRender()` uses `UnifiedDiffOutputBuilder('', TRUE)`, which never emits `\ No newline at end of file` markers, so the trailing-newline state is absent from the diff and `Patcher`'s `implode("\n", ...)` re-derives it from the leftover baseline lines. Repro: baseline `f.txt` = `"old\n"`, actual `f.txt` = `"newval"` (no newline) -> after diff+patch the result is `"newval\n"`. Fix requires emitting and honouring the "no newline" marker end-to-end, which changes the diff-file format.

**B. An added file whose name starts with `-` collides with the deletion-marker convention (medium).** Deletions are encoded as a sibling file whose basename is prefixed with `-`. An added file legitimately named `-report.txt` is copied verbatim by `diff()`, then `patch()` reads the leading `-` as a deletion marker: the added file is dropped, and `-keep.txt` would even delete a real `keep.txt`. Fix requires escaping the marker (a format change).

**C. `CONTENT_IGNORED_MARKER` sentinel can collide with real content (low).** "Ignore content" is stored as the sentinel string `content_ignored` inside `$content`, and `isIgnoreContent()` compares content to it - so a real file whose bytes are exactly `content_ignored` is mis-detected as ignore-content. The clean fix (a dedicated boolean) would break the tested public contract that setting content to `IndexedFile::CONTENT_IGNORED_MARKER` marks a file ignore-content (`IndexedFileTest` asserts exactly this), so it needs a contract decision.

**D. `isSameContent()` inode shortcut ignores the device id (low).** Two same-size files on different filesystems with a coincidentally equal inode are reported identical without hashing. `SplFileInfo` exposes no `st_dev`, so a sound fix means adding a `stat()` (or dropping the inode optimization) - a change to the comparison core's performance profile.

## Verification

- **Tests**: `composer test` -> **244 tests, 678 assertions, OK** (4 new regression tests; green after every commit).
- **Lint**: `composer lint` -> PHPCS (44 files), PHPStan level 9, Rector all clean.
- **Performance**: measured by the CI benchmark job against the committed baseline; the ±5% assertion passes on speedups, so these changes will not trip it.

## Before / After

```
 Comparer::compare() filter cache
 ┌───────────────────────────────────────────────────────────────┐
 │ BEFORE: for each file -> addLeftFile()/addRightFile()          │
 │           -> $this->cache = []   (N resets, path recomputed)   │
 ├───────────────────────────────────────────────────────────────┤
 │ AFTER:  for each file -> set diff on the index key directly    │
 │         -> $this->cache = []     (1 reset, after the loop)     │
 └───────────────────────────────────────────────────────────────┘

 Index::getFiles($cb)
 ┌───────────────────────────────────────────────────────────────┐
 │ BEFORE:  $this->files[$path] = $cb($file);   // cache mutated  │
 │          return $this->files;                // now corrupt    │
 ├───────────────────────────────────────────────────────────────┤
 │ AFTER:   $copy[$path] = $cb($file);          // cache intact   │
 │          return $copy;                       // transformed    │
 └───────────────────────────────────────────────────────────────┘

 Diff::render() on a one-sided diff (added file)
 ┌───────────────────────────────────────────────────────────────┐
 │ BEFORE:  getLeft()->getContent()  ->  \Error: must not be      │
 │          accessed before initialization          (uncatchable) │
 ├───────────────────────────────────────────────────────────────┤
 │ AFTER:   existsLeft() ? ... : ''  ->  clean "+file added"      │
 │          unified diff output                                   │
 └───────────────────────────────────────────────────────────────┘
```

## Summary

- Optimized scan, compare, sync, and patch paths while preserving the public API.
- Added pathname memoization and pattern deduplication.
- Reduced comparison cache resets and clarified patch-file handling.
- Fixed cached-index corruption from callback-based file retrieval.
- Fixed rendering of one-sided diffs.
- Prevented warnings from lone negation lines in `.ignorecontent`.
- Added regression tests for the fixes and edge cases.

## Validation

- 244 tests and 678 assertions passing
- Lint, static analysis, Rector, and benchmark checks passing

## Possibly related

- Possibly related to `#1` (performance benchmarks).

### CODE REVIEW

File: src/Compare/Comparer.php [hunk 1]
```diff
@@ -47,21 +47,27 @@ public function compare(): static {
...
-    foreach ($right_files as $right_file) {
-      $this->addRightFile($right_file);
+    foreach ($right_files as $path => $right_file) {
+      ($this->diffs[$path] ??= new Diff())->setRight($right_file);
     }
 
+    // Filter results derive from $this->diffs; reset the cache once after the
+    // full rebuild instead of on every added file.
+    $this->cache = [];
+
```
[the remark this answers is not in the document]
Comment (AlexSkrypnyk) [reply src/Compare/Comparer.php:73]: Fixed in 28dbfc5. compare() now clears the diffs collection at the start, so a reused Comparer performs a true rebuild without retaining stale entries. Good catch - my comment said 'full rebuild' but the code was merging into prior state.

File: tests/phpunit/Unit/IndexTest.php [hunk 2]
```diff
@@ -154,6 +154,17 @@ public function testGetFilesCalledTwice(): void {
     $this->assertNotEmpty($result1);
   }
 
+  public function testGetFilesWithCallbackDoesNotMutateIndex(): void {
+    $dir = File::dir($this->locationsFixtureDir('compare') . DIRECTORY_SEPARATOR . 'files_equal_advanced' . DIRECTORY_SEPARATOR . 'directory2');
+
+    $index = new Index($dir);
+
+    // Transforming through the callback must not corrupt the cached index.
+    $index->getFiles(fn(IndexedFile $file): string => 'transformed');
```
[the remark this answers is not in the document]
Comment (AlexSkrypnyk) [reply tests/phpunit/Unit/IndexTest.php]: Fixed in 28dbfc5. The callback now derives its return value from the file via getFilename(), removing the unused-parameter warning.
